src/histogram.rs

```diff
@@ -151,16 +154,176 @@ impl From<Opts> for HistogramOpts {
     }
 }
 
+/// Representation of a hot or cold shard.
+///
+/// See [`HistogramCore`] for details.
+#[derive(Debug)]
+struct Shard {
+    sum: AtomicF64,
+    count: AtomicU64,
+    buckets: Vec<AtomicU64>,
+}
+
+impl Shard {
+    fn new(num_buckets: usize) -> Self {
+        let mut buckets = Vec::new();
+        for _ in 0..num_buckets {
+            buckets.push(AtomicU64::new(0));
+        }
+
+        Shard {
+            sum: AtomicF64::new(0.0),
+            count: AtomicU64::new(0),
+            buckets,
+        }
+    }
+}
+
+/// Index into an array of [`Shard`]s.
+///
+/// Used in conjunction with [`ShardAndCount`] below.
+#[derive(Debug, Clone, Copy)]
+pub enum ShardIndex {
+    /// First index. Corresponds to 0.
+    First,
+    /// Second index. Corresponds to 1.
+    Second,
+}
+
+impl ShardIndex {
+    /// Inverse the given [`ShardIndex`].
+    pub fn inverse(self) -> ShardIndex {
+        match self {
+            ShardIndex::First => ShardIndex::Second,
+            ShardIndex::Second => ShardIndex::First,
+        }
+    }
+}
+
+impl From<u64> for ShardIndex {
+    fn from(index: u64) -> Self {
+        match index {
+            0 => ShardIndex::First,
+            1 => ShardIndex::Second,
+            _ => panic!(
+                "Invalid shard index {:?}. A histogram only has two shards.",
+                index
+            ),
+        }
+    }
+}
+
+impl From<ShardIndex> for usize {
+    fn from(index: ShardIndex) -> Self {
+        match index {
+            ShardIndex::First => 0,
+            ShardIndex::Second => 1,
+        }
+    }
+}
+
+/// An atomic u64 with the most significant used as a [`ShardIndex`] and the
+/// remaining 63 bits used to count [`Histogram`] observations.
+#[derive(Debug)]
+pub struct ShardAndCount {
```

Comment:
Does this need to be `pub` or is perhaps `pub(crate)` enough?

Reply:
No need for being public at all. Good catch, thanks!